Normalize x-axis across TopN subcharts

jbcrail opened the pull request:
This PR normalizes the x-axis across all series subcharts in the stacktrace UI as described in https://github.com/elastic/prodfiler/issues/2362.

I also added stronger typing and refactored the server so that the API response requires minimal post-processing by the client.

### Example

The following is a side-by-side comparison of the TopN subcharts before and after normalization of the x-axis.

Before | After
:-------------------------:|:-------------------------:
![before](https://user-images.githubusercontent.com/6038/175608443-3ec115e5-5919-452c-8bd8-ccd59ea3b139.png)  |  ![after](https://user-images.githubusercontent.com/6038/175608456-578ff060-aed9-48d5-b391-0c9a5ce1eb44.png)

jbcrail committed:
Normalize TopN charts

rockdaboot commented on src/plugins/profiling/common/topn.ts:
```diff
@@ -13,42 +13,57 @@ import {
+
+  // Sort by timestamp ascending, count descending, and category ascending
+  samples.sort(
+    (a, b) => a.Timestamp - b.Timestamp || b.Count - a.Count || a.Category.localeCompare(b.Category)
```
Man, I was really puzzled by the `||` operator here 😆

jbcrail replied:
I replaced the multi-term boolean expression with explicit `if` statements.

rockdaboot reviewed (approved):
LGTM

This PR is a good example for where "before" and "after" screenshots are super helpful for the reviewer.

jbcrail commented:
Good point about the screenshots. I added side-by-side screenshots to demonstrate the difference.

jbcrail committed:
Update sort comparator to be more explicit

jbcrail committed:
Fix whitespace

jbcrail committed:
Fix lint warning

jbcrail committed:
Group method with other related methods

rockdaboot commented:
~One more question :-)
On the "after" screenshot, place 3 is "crazy_torvalds". It seems to have much less samples than place 4 and 5 for example. How do you explain this ?~

My bad, #82 fixes it 😃

inge4pres commented on src/plugins/profiling/common/topn.ts:
```diff
@@ -13,42 +13,74 @@ import {
 
-type TopNBucketsByDate = {
-  TopN: Record<number, TopNBucket[]>;
+export type TopNSamples = {
```
I am bit confused on the choice of using a type: if it were an `interface` could we avoid to define type aliases below for all of the variants (`TopNContainers`, `TopNDeployments`, etc...)?

jbcrail replied:
I think the original purpose was to create a union type for a response that either has a `Metadata` field (top N traces) or doesn't have the `Metadata` field (everything else). The type aliases were needed to satisfy the type system.

Since the client-server API is still in flux and a later refactor may require a different design, I switched to interfaces and removed the now-unnecessary type aliases and union type. We may still eventually need to create a response type to accomplish the original purpose, but for now it's not needed.

inge4pres commented on src/plugins/profiling/common/topn.ts:
```diff
@@ -13,42 +13,74 @@ import {
+  return samples;
+}
+
+export function groupSamplesByCategory(samples: TopNSample[]) {
```
My understanding is that we will always want the data returned by `createTopNSamples` to be grouped?
If that is the case, should we inline the grouping logic there and return a tuple / object with both samples and series?

Feel free to ignore this comment if my assumption is wrong, I'm getting back into the Kibana plugin code after a month so my understanding of it needs to be refreshed 😄

jbcrail replied:
Your assumption is right that we do eventually group the `createTopNSamples` data by category. However, since the TopN API response is used by two different charts in the stacktraces UI (i.e. the stacked bar chart and multiple single-series bar charts), I decided to have the API return the common data between the two charts so that the UI can make the other necessary changes as needed:

* The stacked bar chart requires a list of objects, where each object has three fields (timestamp, count, category). In addition, objects with a count of zero are not given to the stacked bar chart since the default tooltip will be cluttered with zero values.

* The multiple single-series bar charts require a list of objects, where each object has two fields (timestamp, count). This is where we have to group by category so that we can dynamically create the individual bar charts for each category. Also, object with zero counts are included so that the x-axis is normalized for all charts.

inge4pres reviewed (approved):
LGTM! 👍🏼 
Thanks for this improvement 🙏🏼 
Two non-blocking comments

dgieselaar commented on src/plugins/profiling/common/topn.ts:
```diff
@@ -13,42 +13,74 @@ import {
+      counts.set(item.key, item.count.value);
     });
+    buckets.set(histogramBuckets[i].key, counts);
   }
```
This might be simplified by:

```ts
const counts = Object.fromEntries(histogram?.buckets.map(bucket => [ bucket.key, bucket.count.value ]);
const uniqueCategories = Object.keys(counts);
```

jbcrail replied:
Unfortunately a few constraints made it difficult to simplify the existing code:

1. `histogram?.buckets` can be undefined so I needed to keep the nullish coalescing operator to avoid using `try/catch`
2. We need to extract three values from `histogram?.buckets` (timestamp, count, category). To do this, we have to follow `histogram?.buckets` down two levels to get all required values. The code using `for` loops versus a functional approach was roughly the same length and complexity.

dgieselaar commented on src/plugins/profiling/common/topn.ts:
```diff
@@ -13,42 +13,74 @@ import {
-  return { TopN: topNBucketsByDate };
+  // Normalize samples so there are an equal number of data points per each timestamp
+  const samples: TopNSample[] = [];
+  for (const timestamp of buckets.keys()) {
+    for (const category of uniqueCategories.values()) {
+      const sample: TopNSample = { Timestamp: timestamp, Count: 0, Category: category };
+      if (buckets.get(timestamp).has(category)) {
+        sample.Count = buckets.get(timestamp).get(category);
+      }
+      samples.push(sample);
+    }
+  }
```
IMHO this is a little opaque - I think renaming some of the variables to signify their meaning might help. E.g., change `counts` to something like `countsByFrameID` (not sure if that is actually what its index is) and `buckets` to `bucketsByTimestamp`.

jbcrail replied:
Agreed, I renamed the suggested variables.

dgieselaar commented on src/plugins/profiling/common/topn.ts:
```diff
@@ -13,42 +13,74 @@ import {
+  }
+
+  // Sort by timestamp ascending, count descending, and category ascending
+  samples.sort((a, b) => {
+    if (a.Timestamp < b.Timestamp) return -1;
+    if (a.Timestamp > b.Timestamp) return 1;
+    if (a.Count > b.Count) return -1;
+    if (a.Count < b.Count) return 1;
+    return a.Category.localeCompare(b.Category);
+  });
+
+  return samples;
+}
```
I'd generally avoid `sort` as it sorts the array in place, similar to `revert()`. It's not an issue here because you're creating a new array a few lines before, but it's easy to gloss over. Additionally, using something like `lodash.orderBy(samples, [ 'Timestamp', 'Count', 'Category' ])` might be more descriptive.

jbcrail replied:
Thanks for the suggestion about lodash. I used `orderBy` and found it to be clearer, especially when sorting numeric fields.

jbcrail committed:
Switch from type to interface

I also removed the type alias for the union type since it's not needed
at this time. The client-server API is still in flux and a later
refactor may require a different design.

jbcrail committed:
Use correct stack accessor for bar chart

jbcrail committed:
Simplify sample ordering

jbcrail committed:
Rename variables for clarity

jbcrail committed:
Refactor sample normalization

jbcrail committed:
Switch from forEach to for loop

This reduce time taken to generate a response.
